snowfall/training/asg_graph.py

```diff
@@ -0,0 +1,128 @@
+# Copyright (c)  2020  Xiaomi Corp.       (author: Fangjun Kuang)
+
+from functools import lru_cache
+from typing import Iterable
+from typing import List
+
+import re
+
+import k2
+import torch
+
+from .ctc_graph import build_ctc_topo
+
+
+def filter_disambig_symbols(symbol_table: k2.SymbolTable,
+                            pattern: str = '#') -> List[int]:
+    '''Return a list of phone IDs containing no disambiguation symbols.
+
+    Caution:
+      You may need to remove the ID with value 0 if the return value
+      is used to create a phone LM.
+
+    Args:
+      symbol_table:
+        A symbol table in k2.
+      pattern:
+        Symbols containing this pattern are disambiguation symbols.
+    Returns:
+      Return a list of symbol IDs excluding those from disambiguation symbols.
+    '''
+    regex = re.compile(pattern)
+    symbols = symbol_table.symbols
+    ans = []
+    for s in symbols:
+        if not regex.search(s):
+            ans.append(symbol_table[s])
+    return ans
+
+
+def create_bigram_phone_lm(phones: List[int]) -> k2.Fsa:
+    '''Create a bigram phone LM.
+    The resulting FSA (P) has a start-state and a state for
+    each phone 0, 1, ....; and each of the above-mentioned states
```

Comment:
I don't think 0 is a phone, it's epsilon I think.

Reply:
You're right. Will change the doc.